rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,700 @@
+
+# User Benefit
+
+## TFX End Users
+
+While this change is transparent to end users, it will facilitate the design and
+implementation of many user-facing features, for example:
+
+*   Columnar storage format in TFX.
+*   Structured training examples.
+
+## Individual TFX component users
+
+We use TFXIO to refer to the proposed I/O abstraction layer. All TFX components
+will start using TFXIO to ingest the data and have a unified way of representing
+the data. Individual TFX component users would be able to implement TFXIO for
+their own data formats / storage formats that are not supported by TFX. By
+design, any such implementation will be readily accessible by all TFX
+components.
+
+## TFX developers
+
+Developers working on TFX infrastructure will not have to understand the
+internals of each component any more in order to make changes to I/O and parsing
+(for example, adding support for a new storage format for the training
+examples).
+
+Developers working on TFX components would benefit from sharing common
+operations against the unified in-memory representation, or even higher-level
+computations. For instance, suppose that we implement a sketch-based algorithm
+to compute approximate heavy hitters over this in-memory representation. We can
+now share this implementation inside both TFDV and TFT for their top-K feature
+value computation.
+
+# Design Proposal
+
+This design proposes **a common in-memory data representation**, **a way to
+translate that into TF feedables** (np.ndarray or EagerTensors) and **a set of
+APIs** each component can use to get both.
+
+![alt_text](20191017-tfx-standardized-inputs/overview.png)
+
+## Common in-memory data representation
+
+[Apache Arrow](https://arrow.apache.org/) will be used as the common in-memory
+data representation. Beam-based TFX components will accept
+<code>PCollection[pyarrow.[RecordBatch](https://arrow.apache.org/docs/python/data.html#record-batches)]</code>.
+
+Each logical data format will have its own encoding convention,
+[discussed](#logical-data-encoding-in-arrow) in the detailed design.
+
+We chose Apache Arrow because:
+
+*   It’s Expressive enough.
+    *   Lossless encoding of (conformant) tf.Example, tf.SequenceExample
+    *   Can encode structured data (proto)
+*   It’s a columnar format. It works well with common TFX workloads:
+    *   Column (feature)-wise analysis
+    *   Feed a batch of columns (features) to TensorFlow.
+*   It’s OSS friendly.
+    *   Community support for more storage format I/O (e.g. Apache Parquet)
+    *   Friendly to other OSS data formats, both in-memory and on disk (e.g.
+        Pandas)
+    *   Friendly to numpy / TF: many Arrow array types share the same memory
+        layout with numpy ndarrays and certain type of TF (composite) Tensors.
+*   TF neutral.
+    *   Leaves the possibility of supporting other ML libraries open.
+
+## Translation from Arrow to TF feedables
```

Comment:
Will it be possible to translate any Arrow `RecordBatch` to a TF representation? e.g. Will we need to do anything special for timestamps, unions, etc.?

Reply:
It's only possible for a subset of Arrow types (listed in this doc, and does not include timestampls or unions).

StructuredTensor was designed to mirror arrow's StructArray which allows nested structured data.